src/analysis/normed_space/dual.lean

```diff
@@ -115,3 +124,152 @@ def inclusion_in_double_dual_li : E →ₗᵢ[𝕜] (dual 𝕜 (dual 𝕜 E)) :=
+lemma of_closed_unit_ball
+  {𝕜 : Type*} [is_R_or_C 𝕜] {E : Type*} [normed_group E] [normed_space 𝕜 E] :
+  polar 𝕜 (closed_ball (0 : E) 1) = {x' : dual 𝕜 E | ∥ x' ∥ ≤ 1 } :=
+begin
+  ext x',
+  simp only [mem_closed_ball, mem_set_of_eq, dist_zero_right],
+  split,
+  { intros h,
+    apply continuous_linear_map.op_norm_le_of_ball zero_lt_one zero_le_one,
+    intros z hz,
+    have key := linear_map.bound_of_ball_bound zero_lt_one 1 x'.to_linear_map h z,
+    simp only [continuous_linear_map.to_linear_map_eq_coe,
+               continuous_linear_map.coe_coe, div_one] at key,
+    exact key, },
+  { intros h z hz,
+    simp only [mem_closed_ball, dist_zero_right] at hz,
+    apply (continuous_linear_map.unit_le_op_norm x' z hz).trans h, },
+end
+
+/-- If `s` is a neighborhood of the origin in a normed space `E`, then at any point `z : E`
+there exists a bound for the norms of the values `x' z` of the elements `x' ∈ polar 𝕜 s` of the
+polar of `s`. -/
+lemma eval_bounded_of_nbhd_zero (𝕜 : Type*) [nondiscrete_normed_field 𝕜]
+  {E : Type*} [normed_group E] [normed_space 𝕜 E]
+  {s : set E} (s_nhd : s ∈ 𝓝 (0 : E)) (z : E) :
+  ∃ (r : ℝ), ∀ (x' : dual 𝕜 E), x' ∈ polar 𝕜 s → ∥ x' z ∥ ≤ r :=
+begin
+  have s_absnt : absorbent 𝕜 s := absorbent_nhds_zero s_nhd,
+  rcases s_absnt z with ⟨c, ⟨c_pos, hc⟩⟩,
+  cases normed_field.exists_lt_norm 𝕜 c with a ha,
+  specialize hc a ha.le,
+  have a_norm_pos : 0 < ∥ a ∥ := lt_trans c_pos ha,
+  have a_ne_zero : a ≠ 0 := norm_pos_iff.mp a_norm_pos,
+  have w_in_s : a⁻¹ • z ∈ s,
+  { rcases hc with ⟨ w , ⟨hws, haw⟩⟩,
+    rwa [← haw, ← mul_smul, inv_mul_cancel a_ne_zero, one_smul], },
+  use ∥a∥,
+  intros x' hx',
+  specialize hx' _ w_in_s,
+  simp only [algebra.id.smul_eq_mul, normed_field.norm_mul,
+             continuous_linear_map.map_smul, normed_field.norm_inv] at hx',
+  have key := mul_le_mul (@rfl _ ∥ a ∥).ge hx' _ (norm_nonneg a),
+  rwa [mul_one, ← mul_assoc, mul_inv_cancel (ne_of_gt a_norm_pos), one_mul] at key,
+  apply mul_nonneg _ (norm_nonneg _),
+  simp only [inv_nonneg, norm_nonneg],
+end
+
+/-- If `s` is a neighborhood of the origin in a normed space `E`, then there exists a
+function `r : E → ℝ` such that for all elements `x' ∈ polar 𝕜 s` one has `∥x' z∥ ≤ r(z)`. -/
+lemma finite_values_of_nbhd_zero {s : set E} (s_nhd : s ∈ 𝓝 (0 : E)) :
+  ∃ (r : E → ℝ), ∀ (x' : dual 𝕜 E) (z : E), x' ∈ polar 𝕜 s → ∥ x' z ∥ ≤ r z :=
+begin
+  cases classical.axiom_of_choice (eval_bounded_of_nbhd_zero 𝕜 s_nhd) with r hr,
+  use r,
+  intros x' z,
+  exact hr z x',
+end
+
+/-- Given a neighborhood `s` of the origin in a normed space `E` over `ℝ` or `ℂ`, the dual norms
+of all elements of the polar `polar 𝕜 s` are bounded by a constant. -/
+lemma bounded_of_nbhd_zero {𝕜 : Type*} [is_R_or_C 𝕜]
+  {E : Type*} [normed_group E] [normed_space 𝕜 E] {s : set E} (s_nhd : s ∈ 𝓝 (0 : E)) :
+  ∃ (c : ℝ), ∀ (x' : dual 𝕜 E), x' ∈ polar 𝕜 s → ∥ x' ∥ ≤ c :=
```

Comment:
True! The ones you mention are not currently used. I agree that if one has `bounded_of_nbhd_zero` in the generality of `[nondiscrete_normed_field 𝕜]`, then they are useless. I will remove the unused ones.

Reply:
Also, the standard abbreviation for neighborhoods in mathlib is nhds, so you should probably use that in the lemma name.

Reply:
Actually I already realize that if I were to now rewrite things using `bounded_of_nbhd_zero` (or `bounded_of_nhds_zero` after a name change), then intermediate compactness statements (not the current formulation of Banach-Alaoglu, though) downstream would be in strictly less generality. So I am starting to have doubts: does it make sense to rewrite things and have less generality?

Reply:
Ok, the observation of @sgouezel is correct and relevant; the current choice is not great style. The lemma `finite_values_of_nbhd_zero` is not used, and I removed it. But `bounds_fun` (and therefore indirectly `eval_bounded_of_nhds_zero`) is used in obtaining compactness in the product space `Π (_ : E), 𝕜` (by Tychonoff's theorem). The way it is currently written I can't see an easy way to change in such a way that the generality of `[nondiscrete_normed_field 𝕜] [proper_space 𝕜]` is kept. Making changes to avoid these lemmas and losing generality does not feel right either, although I agree that these lemmas should probably be replaced in the long run.

For this reason, I'd like to keep the lemmas except from the one that I already deleted.

Reply:
I agree that we should strive for most generality, so my idea would rather be to extend all lemmas needing `is_R_or_C 𝕜` to the general situation of a nondiscrete normed field. Let me have a go at this.

Reply:
I have adapted `bounded_of_nbhd_zero`. Can you update the PR to take this into account?

Reply:
Thank you so much @sgouezel!

When I started writing Banach-Alaoglu, I tried to keep this generality (plus properness assumption for compactness statements), but got badly stuck at this place (probably due to the fact that I have no experience with general nondiscrete normed fields). So I gave up and only did `[is_R_or_C 𝕜]`.

I will of course adapt the rest of the two PRs to take into account this great improvement. Thank you again!

Reply:
I adapted the two PRs. Banach-Alaoglu theorem about the weak-star compactness of polar sets of neighborhoods of the origin is now proved in the generality where the ground field `𝕜` satisfies `[nondiscrete_normed_field 𝕜] [proper_space 𝕜]`. This was made possible by Sébastien's improvement.

I entirely removed the old `def polar.bounds_fun` and `lemma polar.bounds_fun_spec` which assumed `is_R_or_C 𝕜]`.

In the most straightforward adaptation (https://github.com/leanprover-community/mathlib/pull/9836/commits/5482fe32bb86029224e4d8d022a34b5270c3f564), I did correspondingly create `def polar.bound_of_nhds_zero` and a few lemmas about it. This new definition and lemmas did not seem useful beyond the Banach-Alaoglu purpose, although at least they are not in the right generality. They were used in `pi_ball_bounds_fun_cpt`, `image_polar_nhd_subset`, and `image_polar_nhd_compact` in the follow-up PR #9862.

My current thinking is that the cleanest is to remove these and remove `pi_ball_bounds_fun_cpt`, `image_polar_nhd_subset`, at the expense of making the proof of `image_polar_nhd_compact` a bit longer. This is my current proposal. If this seems right, I will update the follow-up PR #9862 accordingly.